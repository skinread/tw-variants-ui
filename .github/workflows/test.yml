# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
