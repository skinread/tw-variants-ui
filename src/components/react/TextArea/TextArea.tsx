import { forwardRef, useRef } from 'react';
import { inputVariants, validityFromProps } from '@common';
import { type CommonFieldProps } from '@components/react';

type TextAreaPropsFiltered = Omit<React.ComponentPropsWithoutRef<'textarea'>, 'name'>;

export interface TextAreaProps
  extends Pick<CommonFieldProps, 'label' | 'name' | 'hasError' | 'id'>,
    TextAreaPropsFiltered {
  /**
   * The placeholder text will appear in the field while it is empty to help instruct the user
   */
  placeholder?: string;
}

/**
 * Styled text area input field. React `ref` enabled.
 * Also exposes intrinsic `<textarea>` attributes.
 */
export const TextArea = forwardRef<HTMLTextAreaElement, TextAreaProps>(
  function VuiTextArea(props, ref) {
    const { hasError, id: idProp, label, ...elemAttrs } = props;
    const {
      wrapper: wrapperStyle,
      label: labelStyle,
      labelText: labelTextStyle,
      input: inputStyle,
    } = inputVariants({ type: 'textarea', validity: validityFromProps({ hasError }) });
    const id = useRef(idProp ?? `text-area-${props.name}`);

    return (
      <div className={wrapperStyle()}>
        <label className={labelStyle()} htmlFor={id.current}>
          <span className={labelTextStyle()}>{label}</span>
        </label>
        <textarea id={id.current} className={inputStyle()} {...elemAttrs} ref={ref}></textarea>
      </div>
    );
  }
);
