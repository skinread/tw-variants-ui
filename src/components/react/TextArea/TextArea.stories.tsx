import type { Meta, StoryObj } from '@storybook/react';
import { expect, within, userEvent } from '@storybook/test';

import { TextArea } from './TextArea';

const meta = {
  title: 'React/Text Area',
  component: TextArea,
  tags: ['autodocs'],
  args: {
    label: 'Tell me a story',
    name: 'demo-textarea-field',
    placeholder: 'Once upon a time...',
  },
  argTypes: {
    onChange: { action: 'change' },
    onFocus: { action: 'focus' },
    onBlur: { action: 'blur' },
  },
} satisfies Meta<typeof TextArea>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic: Story = {};

export const InteractionTest: Story = {
  args: {
    name: 'interaction-test',
  },
  play: async ({ args, canvasElement, step }) => {
    const canvas = within(canvasElement);
    const field = canvas.getByRole('textbox');

    await step('Initial state', async () => {
      await expect(field).toHaveAttribute('name', args.name);
      await expect(field).toHaveAttribute('id', `text-area-${args.name}`);
      await expect(field).toHaveValue('');
    });

    await step('Focus on field', async () => {
      await userEvent.click(field);
      await expect(args.onFocus).toHaveBeenCalled();
    });

    await step('Sample input', async () => {
      const content = 'Hello, happy testing!';
      await userEvent.type(field, content);
      await expect(args.onChange).toHaveBeenCalledTimes(content.length);
    });

    await step('Remove focus from field', async () => {
      await userEvent.keyboard('[Tab]');
      await expect(args.onBlur).toHaveBeenCalled();
    });
  },
};
