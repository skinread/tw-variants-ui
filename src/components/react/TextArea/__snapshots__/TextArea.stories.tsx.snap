// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Text Area Basic smoke-test 1`] = `
<div class="form-control">
  <label class="label items-start px-0"
         for="text-area-demo-textarea-field"
  >
    <span class="label-text text-base font-bold">
      Tell me a story
    </span>
  </label>
  <textarea id="text-area-demo-textarea-field"
            class="placeholder:text-subtle textarea textarea-primary text-base rounded-[--rounded-field]"
            name="demo-textarea-field"
            placeholder="Once upon a time..."
  >
  </textarea>
</div>
`;
